# Exercise 4.08 - Grouping Data & Aggregating Variables

In [4]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [6]:
# create path
path='/home/haus/Exercise_4_Instacart_Basket_Analysis/02_Data'

In [8]:
# Importing ords_prods_merge
ords_prods_merge=pd.read_pickle(r'/home/haus/Exercise_4_Instacart_Basket_Analysis/02_Data/Prepared_Data/orders_products_merged_derived.pkl')

In [9]:
# Creating a subset of the data-frame "ords_prods_merge"
df_subset = ords_prods_merge[:1000000]

In [10]:
# Checking the subset shape
df_subset.shape

(1000000, 21)

In [11]:
# Checking the subset head
df_subset.head()

,order_id,user_id,eval_set,order_number,order_dow,order_hour_of_day,days_since_prior_order,first_order,product_id,add_to_cart_order,...,Unnamed: 0_y,product_name,aisle_id,department_id,prices,merge_info,price_range_loc,busiest_day,Busiest_days,busiest_period_of_day
0,2539329,1,prior,1,2,8,NaN,True,196,1,...,195,Soda,77,7,9.0,both,Mid-range product,Regularly busy,Regularly busy days,Average orders
1,2539329,1,prior,1,2,8,NaN,True,14084,2,...,14084,Organic Unsweetened Vanilla Almond Milk,91,16,12.5,both,Mid-range product,Regularly busy,Regularly busy days,Average orders
2,2539329,1,prior,1,2,8,NaN,True,12427,3,...,12427,Original Beef Jerky,23,19,4.4,both,Low-range product,Regularly busy,Regularly busy days,Average orders
3,2539329,1,prior,1,2,8,NaN,True,26088,4,...,26089,Aged White Cheddar Popcorn,23,19,4.7,both,Low-range product,Regularly busy,Regularly busy days,Average orders
4,2539329,1,prior,1,2,8,NaN,True,26405,5,...,26406,XL Pick-A-Size Paper Towel Rolls,54,17,1.0,both,Low-range product,Regularly busy,Regularly busy days,Average orders


In [12]:
# Grouping data with Pandas
df_subset.groupby('product_name')

In [18]:
# How to produce a single descriptive statistic for the column “order_number”
df_subset.groupby('department_id').agg({'order_number': ['mean']})

,order_number
,mean
department_id,
1,14.800024
2,17.091743
3,17.913544
4,17.893092
5,15.214270
6,15.382135
7,17.694027
8,16.458105


In [20]:
# Ho to produce multiple statistics at the same time by adding more arguments to your code:
df_subset.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                14.800024   1  99
2                17.091743   1  98
3                17.913544   1  99
4                17.893092   1  99
5                15.214270   1  99
6                15.382135   1  99
7                17.694027   1  99
8                16.458105   1  91
9                15.957363   1  99
10               20.091818   1  99
11               16.482026   1  99
12               15.615061   1  99
13               16.484023   1  99
14               17.524632   1  99
15               15.691875   1  99
16               18.014071   1  99
17               16.150593   1  99
18               19.602850   1  99
19               17.631340   1  99
20               17.138607   1  99
21               21.956893   1  97

In [22]:
# How to create a column named "max_order" in your data-frame (ords_prods_merge)
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/tmp/ipykernel_91109/2155634612.py:2: FutureWarning: The provided callable <function max at 0x75f6b20ff060> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


# there is a warning message:
It means that while np.max currently maps to the behavior of SeriesGroupBy.max, future versions of pandas will interpret it differently. To ensure compatibility with future versions and avoid this warning, you should replace np.max with the string "max".
# Modified code to avoid the warning:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform('max')

In [25]:
# Checking the data-frame after having created the new column
ords_prods_merge.head(15)

,order_id,user_id,eval_set,order_number,order_dow,order_hour_of_day,days_since_prior_order,first_order,product_id,add_to_cart_order,...,product_name,aisle_id,department_id,prices,merge_info,price_range_loc,busiest_day,Busiest_days,busiest_period_of_day,max_order
0,2539329,1,prior,1,2,8,NaN,True,196,1,...,Soda,77,7,9.0,both,Mid-range product,Regularly busy,Regularly busy days,Average orders,10
1,2539329,1,prior,1,2,8,NaN,True,14084,2,...,Organic Unsweetened Vanilla Almond Milk,91,16,12.5,both,Mid-range product,Regularly busy,Regularly busy days,Average orders,10
2,2539329,1,prior,1,2,8,NaN,True,12427,3,...,Original Beef Jerky,23,19,4.4,both,Low-range product,Regularly busy,Regularly busy days,Average orders,10
3,2539329,1,prior,1,2,8,NaN,True,26088,4,...,Aged White Cheddar Popcorn,23,19,4.7,both,Low-range product,Regularly busy,Regularly busy days,Average orders,10
4,2539329,1,prior,1,2,8,NaN,True,26405,5,...,XL Pick-A-Size Paper Towel Rolls,54,17,1.0,both,Low-range product,Regularly busy,Regularly busy days,Average orders,10
5,2398795,1,prior,2,3,7,15.0,False,196,1,...,Soda,77,7,9.0,both,Mid-range product,Regularly busy,Lowest Busy days,Average orders,10
6,2398795,1,prior,2,3,7,15.0,False,10258,2,...,Pistachios,117,19,3.0,both,Low-range product,Regularly busy,Lowest Busy days,Average orders,10
7,2398795,1,prior,2,3,7,15.0,False,12427,3,...,Original Beef Jerky,23,19,4.4,both,Low-range product,Regularly busy,Lowest Busy days,Average orders,10
8,2398795,1,prior,2,3,7,15.0,False,13176,4,...,Bag of Organic Bananas,24,4,10.3,both,Mid-range product,Regularly busy,Lowest Busy days,Average orders,10
9,2398795,1,prior,2,3,7,15.0,False,26088,5,...,Aged White Cheddar Popcorn,23,19,4.7,both,Low-range product,Regularly busy,Lowest Busy days,Average orders,10


Create a flag to assigns a “loyalty” label to a user ID based on its corresponding max order value

In [27]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [29]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [31]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [33]:
# Check the frequency of the new column “loyalty_flag”
ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [35]:
# Checking the data-frame after having created the new column
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

,user_id,loyalty_flag,order_number
0,1,New customer,1
1,1,New customer,1
2,1,New customer,1
3,1,New customer,1
4,1,New customer,1
5,1,New customer,2
6,1,New customer,2
7,1,New customer,2
8,1,New customer,2
9,1,New customer,2


Step 2 - Find the aggregated mean of the “order_number” column grouped by “department_id” for the entire dataframe.

In [37]:
# How to produce a single descriptive statistic for the column “order_number”
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

,order_number
,mean
department_id,
1,15.457838
2,17.277920
3,17.170395
4,17.811403
5,15.215751
6,16.439806
7,17.225802
8,15.340650


Step3 - How do the results for the entire dataframe differ from those of the subset?
Compared to the results from the subset, the results from the entire dataframe includes all 'department_id' values.
Additionally, the mean values for the department ids present in the subset have all changed. Some have increased while others decreased.

Step 4 - Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.
This task has been excuted on the data-frame during the exercise (please see the cells above)

Step 5 - The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [39]:
# Check wehther the prices of products purchased by loyal customers differ from those purchased by regular or new customer
ords_prods_merge.groupby(['loyalty_flag']).agg({'prices':['mean','min','max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773487  1.0  25.0
New customer      7.801077  1.0  25.0
Regular customer  7.798145  1.0  25.0

Observations
Mean Prices:
The mean price for loyal customers (7.773487) is slightly lower than that for new customers (7.801077) and regular customers (7.798145).
However, the difference in mean prices between these groups is very small (less than 0.03).
Minimum and Maximum Prices:
The minimum price (1.0) and maximum price (25.0) are identical across all loyalty groups.
This suggests that all customer groups have access to the same range of product prices.

Conclusion:
The prices of products purchased by loyal customers do not differ significantly from those purchased by regular or new customers:
The mean price differences are negligible, indicating that loyalty status does not strongly influence the average price of products purchased.
The identical minimum and maximum prices suggest that all groups are purchasing from the same price range.

Step 6 - Create a spending flag for each user based on the average price across all their orders using the following criteria:
If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [41]:
# Calculating average spend per user
ords_prods_merge['average_spend'] = ords_prods_merge.groupby('user_id')['prices'].transform('mean')

In [43]:
# Creating the flag for Low_spender customers
ords_prods_merge.loc[ords_prods_merge['average_spend'] < 10, 'spending_flag'] = 'Low_spender'

In [45]:
# Creating the flag for High_spend customers
ords_prods_merge.loc[ords_prods_merge['average_spend'] >= 10, 'spending_flag'] = 'High_spender'

In [47]:
# Check customers spender flags
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low_spender     32285150
High_spender      119709
Name: count, dtype: int64

Step 7 - Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the days_since_prior_order” column.
The criteria for the flag should be as follows:
If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [49]:
# Create a column for frequent customers/users
ords_prods_merge['Customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [51]:
# Creating the flag for condition 1 (Non-frequent customer)
ords_prods_merge.loc[ords_prods_merge['Customer_frequency'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [53]:
# Creating the flag for condition 2 (Regular customer)
ords_prods_merge.loc[(ords_prods_merge['Customer_frequency'] > 10) & (ords_prods_merge['Customer_frequency'] <= 20), 'frequency_flag'] = 'Regular customer'

In [55]:
# creating the flag for condition 3 (Frequent customer)
ords_prods_merge.loc[ords_prods_merge['Customer_frequency'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [57]:
# Check customers frequency flag
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

The NaN values could mean that there are customers who have only made one purchase.
This should be investigated for a better understanding.

In [59]:
# Checking the data-frame after having created the new column
ords_prods_merge[['user_id','days_since_prior_order', 'Customer_frequency', 'frequency_flag']].head(100)

,user_id,days_since_prior_order,Customer_frequency,frequency_flag
0,1,NaN,20.5,Non-frequent customer
1,1,NaN,20.5,Non-frequent customer
2,1,NaN,20.5,Non-frequent customer
3,1,NaN,20.5,Non-frequent customer
4,1,NaN,20.5,Non-frequent customer
...,...,...,...,...
95,2,8.0,13.0,Regular customer
96,2,8.0,13.0,Regular customer
97,2,8.0,13.0,Regular customer
98,2,8.0,13.0,Regular customer


Step 9 - Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder

In [61]:
ords_prods_merge.to_pickle(os.path.join(path, 'Prepared_Data', 'aggregated_ords_prods_merge.pkl'))

Close

Steps to Verify the hypothesis related to NaN values:
Count the Number of Unique Orders per User:
Group by user_id and count the unique values in the order_number column.
Identify Users with Only One Order:
Filter users who have only one unique order.
Compare with Users Having NaN in frequency_flag:
Extract users whose frequency_flag is NaN and check if they match those with only one order.

In [64]:
# Count the number of unique orders for each user
user_order_counts = ords_prods_merge.groupby('user_id')['order_number'].nunique()

In [66]:
# Identify users with only one order
single_order_users = user_order_counts[user_order_counts == 1].index

In [68]:
# Check if these users correspond to the NaN values in the frequency_flag column
nan_users = ords_prods_merge[ords_prods_merge['frequency_flag'].isna()]['user_id'].unique()

In [70]:
# Compare the two sets of users
single_order_nan_match = set(single_order_users) == set(nan_users)

In [72]:
# Output the result
print("Do NaN values correspond to single-order users?", single_order_nan_match)
print("Number of users with NaN frequency_flag:", len(nan_users))
print("Number of users with only one order:", len(single_order_users))

Do NaN values correspond to single-order users? False
Number of users with NaN frequency_flag: 1
Number of users with only one order: 2


The results indicate that the NaN values in the frequency_flag column do not correspond exclusively to users with only one order. Specifically:
Do NaN values correspond to single-order users?: False
This means there is no perfect match between the user IDs with NaN in frequency_flag and those with only one order.
Number of users with NaN frequency_flag: 1
Only one user has a NaN value in the frequency_flag column.
Number of users with only one order: 2
Two users have only one order in the dataset.

Possible Causes
Missing or Incomplete Data:
The user with NaN in frequency_flag might have missing values in the days_since_prior_order column, which prevents calculation of the median.
The second single-order user might have valid data for days_since_prior_order, allowing their median to be calculated.
Data Processing Issue:
There might be inconsistencies in how the frequency_flag column was assigned, leading to a mismatch between single-order users and those with NaN.